xls/modules/rle/rle.x

```diff
@@ -0,0 +1,314 @@
+// Copyright 2023 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the symbols from
+// the input stream. The output contains packets of data with the symbols and
+// the number of their occurrences in the input stream.
+//
+// The RLE encoder exposes one input and one output channel. The input channel
+// contains the symbol to compress and flush request. The request is used to
+// dump all the data from the encoder. A typical use of the flush field is
```

Comment:
I wonder if it would make sense to make the "typical use" the "only supported use" in order simplify the protocol, if we only use flash to end (or start the transmission), maybe we could simply ignore the `symbol` part of the packet?

That would require the user to do always send the last `symbol` before the `flash` cycle, and remove the need for saving the `prev_flush` state?

wdyt?

Reply:
Looking at the [AXIS stream protocol specification](https://developer.arm.com/documentation/ihi0051/latest/) (is that the right version to look at?)

It does seem to allow `TLAST` with no accompanying data, see:

> 2.6.1 Transfer with zero data or position bytes
> A transfer can have TLAST asserted but contain no data or position bytes. This can be used to:
> • Indicate the end of a packet when there are no more data or position bytes to transmit.
> • Push through any data that is held in intermediate buffers.
> • Complete an operation at an end-point that is expecting a TLAST at the end of a packet.

Some key differences I see with an empty delimiter packet:
a. the proc is guaranted to receive packet on each tick
b. the receiver can be guaranted that the last `EncOutData` will be emitted in the same clock cycle as `flush`
c. the producer potentially needs to buffer one `symbol` in case the flush needs to happen without interrupting the stream

(a) and (b) sound desirable, (c) seems like a limitation but it seems that it would also happen with the current implementation: as the proc would not attempt to read when "acting" on the `flush` state and I'm not sure how common it would be: naively I would though that the need to flush would be strongly associated with some boundary being detected in the input stream.